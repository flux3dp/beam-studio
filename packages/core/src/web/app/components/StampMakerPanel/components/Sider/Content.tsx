import React, { useMemo } from 'react';

import { Flex, Form, Switch } from 'antd';
import Konva from 'konva';

import UnitInput from '@core/app/widgets/UnitInput';
import useI18n from '@core/helpers/useI18n';
import browser from '@core/implementations/browser';

import { useStampMakerPanelStore } from '../../store';

import styles from './Content.module.scss';

export default function Content(): React.JSX.Element {
  const { stamp_maker_panel } = useI18n();
  const { bevelRadius, filters, horizontalFlip, setBevelRadius, setHorizontalFlip, toggleInvert } =
    useStampMakerPanelStore();
  const isInverted = useMemo(() => filters.includes(Konva.Filters.Invert), [filters]);

  const handleToggleFlip = () => setHorizontalFlip(!horizontalFlip);

  return (
    <div className={styles.wrapper}>
      <Form layout="horizontal">
        <div className={styles.link} onClick={() => browser.open(stamp_maker_panel.how_to.url)}>
          {stamp_maker_panel.how_to.text}
        </div>
        <Form.Item label={stamp_maker_panel.invert}>
          <Flex align="center" gap={8}>
            <Switch checked={isInverted} onChange={toggleInvert} />
          </Flex>
        </Form.Item>
        <Form.Item label={stamp_maker_panel.horizontal_flip}>
          <Flex align="center" gap={8}>
            <Switch checked={horizontalFlip} onChange={handleToggleFlip} />
          </Flex>
        </Form.Item>
        <Form.Item label={stamp_maker_panel.bevel_radius} layout="vertical">
          <Flex justify="space-between">
            <UnitInput
              addonAfter="mm"
              max={10}
              min={0}
              onChange={(value) => {
                if (value !== null) {
                  setBevelRadius(value);
                }
              }}
              precision={1}
              step={0.1}
              value={bevelRadius}
            />
          </Flex>
        </Form.Item>
      </Form>
    </div>
  );
}
