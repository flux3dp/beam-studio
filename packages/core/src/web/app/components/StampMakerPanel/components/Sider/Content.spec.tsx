import React from 'react';
import { fireEvent, render } from '@testing-library/react';

const mockKonvaFilters = {
  Invert: jest.fn(),
};

jest.mock('konva', () => ({
  Filters: mockKonvaFilters,
}));

const mockUseStampMakerPanelStore = jest.fn();
const mockToggleInvert = jest.fn();
const mockSetBevelRadius = jest.fn();
const mockSetHorizontalFlip = jest.fn();

jest.mock('../../store', () => ({
  useStampMakerPanelStore: mockUseStampMakerPanelStore,
}));

jest.mock('@core/app/widgets/UnitInput', () => ({ addonAfter, max, min, onChange, step, value }: any) => (
  <input
    data-addon={addonAfter}
    data-testid="unit-input"
    max={max}
    min={min}
    onChange={(e) => onChange(Number(e.target.value))}
    step={step}
    type="number"
    value={value}
  />
));

import Content from './Content';

describe('test Content', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    mockUseStampMakerPanelStore.mockReturnValue({
      bevelRadius: 0,
      filters: [],
      horizontalFlip: false,
      setBevelRadius: mockSetBevelRadius,
      setHorizontalFlip: mockSetHorizontalFlip,
      toggleInvert: mockToggleInvert,
    });
  });

  it('should render correctly', () => {
    const { container } = render(<Content />);

    expect(container).toMatchSnapshot();
  });

  it('should render invert switch unchecked when filter not applied', () => {
    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const invertSwitch = switches[0]; // First switch is Invert

    expect(invertSwitch.getAttribute('aria-checked')).toBe('false');
  });

  it('should render invert switch checked when filter is applied', () => {
    mockUseStampMakerPanelStore.mockReturnValue({
      bevelRadius: 0,
      filters: [mockKonvaFilters.Invert],
      horizontalFlip: false,
      setBevelRadius: mockSetBevelRadius,
      setHorizontalFlip: mockSetHorizontalFlip,
      toggleInvert: mockToggleInvert,
    });

    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const invertSwitch = switches[0]; // First switch is Invert

    expect(invertSwitch.getAttribute('aria-checked')).toBe('true');
  });

  it('should call toggleInvert when invert switch is clicked', () => {
    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const invertSwitch = switches[0]; // First switch is Invert

    fireEvent.click(invertSwitch);

    expect(mockToggleInvert).toHaveBeenCalled();
  });

  it('should call toggleInvert when invert switch is clicked (when already inverted)', () => {
    mockUseStampMakerPanelStore.mockReturnValue({
      bevelRadius: 0,
      filters: [mockKonvaFilters.Invert],
      horizontalFlip: false,
      setBevelRadius: mockSetBevelRadius,
      setHorizontalFlip: mockSetHorizontalFlip,
      toggleInvert: mockToggleInvert,
    });

    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const invertSwitch = switches[0]; // First switch is Invert

    fireEvent.click(invertSwitch);

    expect(mockToggleInvert).toHaveBeenCalled();
  });

  it('should render horizontal flip switch with correct state', () => {
    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const flipSwitch = switches[1]; // Second switch is Flip Horizontally

    expect(flipSwitch.getAttribute('aria-checked')).toBe('false');
  });

  it('should toggle horizontal flip when switch clicked', () => {
    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const flipSwitch = switches[1]; // Second switch is Flip Horizontally

    fireEvent.click(flipSwitch);

    expect(mockSetHorizontalFlip).toHaveBeenCalledWith(true);
  });

  it('should toggle horizontal flip off when already on', () => {
    mockUseStampMakerPanelStore.mockReturnValue({
      bevelRadius: 0,
      filters: [],
      horizontalFlip: true,
      setBevelRadius: mockSetBevelRadius,
      setHorizontalFlip: mockSetHorizontalFlip,
      toggleInvert: mockToggleInvert,
    });

    const { getAllByRole } = render(<Content />);

    const switches = getAllByRole('switch');
    const flipSwitch = switches[1]; // Second switch is Flip Horizontally

    fireEvent.click(flipSwitch);

    expect(mockSetHorizontalFlip).toHaveBeenCalledWith(false);
  });

  it('should render bevel radius input with correct value', () => {
    mockUseStampMakerPanelStore.mockReturnValue({
      bevelRadius: 2.5,
      filters: [],
      horizontalFlip: false,
      setBevelRadius: mockSetBevelRadius,
      setHorizontalFlip: mockSetHorizontalFlip,
      toggleInvert: mockToggleInvert,
    });

    const { getByTestId } = render(<Content />);

    const bevelInput = getByTestId('unit-input') as HTMLInputElement;

    expect(bevelInput.value).toBe('2.5');
    expect(bevelInput.min).toBe('0');
    expect(bevelInput.max).toBe('10');
    expect(bevelInput.step).toBe('0.1');
    expect(bevelInput.dataset.addon).toBe('mm');
  });

  it('should update bevel radius when input changes', () => {
    const { getByTestId } = render(<Content />);

    const bevelInput = getByTestId('unit-input');

    fireEvent.change(bevelInput, { target: { value: '5' } });

    expect(mockSetBevelRadius).toHaveBeenCalledWith(5);
  });

  it('should not update bevel radius when value is null', () => {
    const { getByTestId } = render(<Content />);

    // Mock UnitInput to simulate null value
    const bevelInput = getByTestId('unit-input');

    fireEvent.change(bevelInput, { target: { value: '' } });

    // Since our mock always returns a number, we need to test the actual component logic
    // In the real component, null values are ignored
    expect(mockSetBevelRadius).not.toHaveBeenCalledWith(null);
  });

  it('should render tooltips for form items', () => {
    const { container } = render(<Content />);

    const tooltipIcons = container.querySelectorAll('.anticon-question-circle');

    expect(tooltipIcons).toHaveLength(2);
  });
});
