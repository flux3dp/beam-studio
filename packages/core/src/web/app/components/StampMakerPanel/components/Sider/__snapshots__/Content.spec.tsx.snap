// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Content should render correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-hash"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-hash"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-hash"
          >
            <label
              class=""
              title="Invert"
            >
              Invert
              <span
                aria-label="question-circle"
                class="anticon anticon-question-circle ant-form-item-tooltip"
                role="img"
                title=""
              >
                <svg
                  aria-hidden="true"
                  data-icon="question-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                  />
                </svg>
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-hash"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="false"
                  class="ant-switch css-dev-only-do-not-override-hash"
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-hash"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-hash"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-hash"
          >
            <label
              class=""
              title="Horizontal Flip"
            >
              Horizontal Flip
              <span
                aria-label="question-circle"
                class="anticon anticon-question-circle ant-form-item-tooltip"
                role="img"
                title=""
              >
                <svg
                  aria-hidden="true"
                  data-icon="question-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                  />
                </svg>
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-hash"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="false"
                  class="ant-switch css-dev-only-do-not-override-hash"
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-hash ant-form-item-vertical"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-hash"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-hash"
          >
            <label
              class=""
              title="Bevel Radius"
            >
              Bevel Radius
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-hash"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-flex css-dev-only-do-not-override-hash ant-flex-justify-space-between"
                >
                  <input
                    data-addon="mm"
                    data-testid="unit-input"
                    max="10"
                    min="0"
                    step="0.1"
                    type="number"
                    value="0"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
